### PR TITLE
Add options to select quality on media remote

### PR DESCRIPTION
**Pull request by jdragonbae**
This patch means adding option to select quality on media remote. It's only for
YouTube site because java script used is defined by the YouTube site.

Refs: #776

### CODE REVIEW

**Review by romandev — COMMENTED** [review]
Great works!
Just query: How do we get the current quality? It is always auto?

**Comment by jdragonbae** [comment]
> Great works!
> Just query: How do we get the current quality? It is always auto?

@romandev Thank you!  Yes, you're right.
 
We need a way to get a current quality. Probably, using javascript query is one of the ways.
`document.querySelector('.html5-video-player').getPlaybackQuality()`, 
`document.querySelector('.html5-video-player').getAvailableQualityLevels()`
But, this way needs javascript callback to get the result. WDYT?

**Review by romandev — COMMENTED** [review]
Okay, let's separate the JavaScript Callback patch from this PR and re-upload it first.

File: src/org/triple/banana/media_remote/MediaRemoteViewImpl.java [hunk 1]
```diff
@@ -181,6 +189,15 @@ private void relayoutContentView() {
                 v.setVisibility(View.VISIBLE);
             });
         }
+
+        BananaTab tab = org.banana.cake.interfaces.BananaTabManager.get().getActivityTab();
+        if (tab == null || tab.getContext() == null) return;
+        if (MediaSuspendController.instance.isYoutubeDomainUrl(tab.getUrl())) {
```
Comment (romandev) [inline src/org/triple/banana/media_remote/MediaRemoteViewImpl.java:195]: This is not related to MediaSuspendController. You should separate the method to common class.

File: src/org/triple/banana/media_remote/MediaRemoteViewImpl.java [hunk 2]
```diff
@@ -206,6 +223,12 @@ private void resetContentViewInteractionListeners() {
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
             $.select(R.id.pip_button, v -> v.setOnClickListener(null));
         }
+
+        BananaTab tab = org.banana.cake.interfaces.BananaTabManager.get().getActivityTab();
```
Comment (romandev) [inline src/org/triple/banana/media_remote/MediaRemoteViewImpl.java:227]: Unnecessary, please add the `R.id.quality_button` to `buttons`

File: src/org/triple/banana/media_remote/MediaRemoteViewImpl.java [hunk 3]
```diff
@@ -22,16 +22,23 @@
 import androidx.annotation.IdRes;
```
Comment (romandev) [inline src/org/triple/banana/media_remote/MediaRemoteViewImpl.java:22]: Please move most of this change in this file to new separated class.

File: src/org/triple/banana/media/MediaCommandProcessor.java [hunk 4]
```diff
@@ -25,6 +25,17 @@ private void runMediaCommand(String mediaCommand) {
         tab.evaluateJavaScript(commandBuilder.toString());
     }
 
+    private void runMediaCommandOnYoutube(String mediaCommand) {
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/media/MediaCommandProcessor.java]: Thanks for the review. I try to separate class from `MediaCommandProcessor`.

File: src/org/triple/banana/media/MediaController.java [hunk 5]
```diff
@@ -42,4 +42,8 @@ public void setVolume(float value) {
     public void setRelativeVolume(float value) {
         mCommandProcessor.setRelativeVolume(value);
     }
+
+    public void setQuality(String quality) {
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/media/MediaController.java]: Ack.

File: src/org/triple/banana/media_remote/MediaRemoteViewImpl.java (hunk 1 above)
Comment (jdragonbae) [reply src/org/triple/banana/media_remote/MediaRemoteViewImpl.java]: Thanks! Ack.

File: src/org/triple/banana/media_remote/MediaRemoteViewImpl.java (hunk 2 above)
Comment (jdragonbae) [reply src/org/triple/banana/media_remote/MediaRemoteViewImpl.java]: Done.

File: src/org/triple/banana/media_remote/MediaRemoteViewImpl.java [hunk 6]
```diff
@@ -181,8 +185,17 @@ private void relayoutContentView() {
                 v.setVisibility(View.VISIBLE);
             });
         }
-    }
 
+        // Quality button is supported only on YouTube site.
+        BananaTab tab = org.banana.cake.interfaces.BananaTabManager.get().getActivityTab();
```
Comment (romandev) [inline src/org/triple/banana/media_remote/MediaRemoteViewImpl.java:190]: Add a new version of `isYoutubeDomainUrl` that queries the activity tab internally.

File: src/org/triple/banana/media_remote/MediaRemoteViewImpl.java [hunk 7]
```diff
@@ -264,6 +280,27 @@ public void onUpdate(MediaRemoteViewModel.ReadonlyData data) {
         updateTimeInfo(data);
     }
 
+    @Override
+    public void showQualityOptionsDialog(int selectedQuality, String[] availableQualites) {
+        if (mDialog == null || mDialog.getContext() == null) return;
+        if (mQualityOptionsDialog == null) {
```
Comment (romandev) [inline src/org/triple/banana/media_remote/MediaRemoteViewImpl.java:286]: Move this code to Youtube util

File: src/org/triple/banana/media_remote/MediaRemoteService.java [hunk 8]
```diff
@@ -157,6 +159,10 @@ public void onMediaRemoteButtonClicked(int id) {
             AudioUtil.setMediaVolumeMuted(!isVolumeMuted);
             mViewModel.getEditor().setIsMuted(!isVolumeMuted);
             mViewModel.commit();
+        } else if (id == R.id.quality_button) {
+            mView.showQualityOptionsDialog(0 /* auto */,
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/media_remote/MediaRemoteService.java]: Done.

File: src/org/triple/banana/media/MediaYoutubeCommander.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    public void setQuality(String quality) {
+        if (!MEDIA_QUALITES.containsKey(quality)) return;
+        quality = MEDIA_QUALITES.get(quality);
+        String script = "node.setPlaybackQualityRange('" + quality + "');"
+                + "node.setPlaybackQuality('" + quality + "');";
+        runCommandOnYoutube(script, null);
+    }
+
+    private static Map<String, String> createMediaQualityMap() {
+        Map<String, String> map = new LinkedHashMap<>();
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/media/MediaYoutubeCommander.java]: Ack.

File: src/org/triple/banana/media_remote/MediaRemoteViewImpl.java (hunk 6 above)
Comment (jdragonbae) [reply src/org/triple/banana/media_remote/MediaRemoteViewImpl.java]: Ack.

File: src/org/triple/banana/media_remote/MediaRemoteViewImpl.java (hunk 7 above)
Comment (jdragonbae) [reply src/org/triple/banana/media_remote/MediaRemoteViewImpl.java]: Done.

File: src/org/triple/banana/media_remote/MediaRemoteViewImpl.java [hunk 10]
```diff
@@ -181,8 +185,17 @@ private void relayoutContentView() {
                 v.setVisibility(View.VISIBLE);
             });
         }
-    }
 
+        // Quality button is supported only on YouTube site.
+        BananaTab tab = org.banana.cake.interfaces.BananaTabManager.get().getActivityTab();
+        if (tab == null || tab.getContext() == null) return;
+        if (UrlUtil.isYoutubeDomainUrl(tab.getUrl())) {
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/media_remote/MediaRemoteViewImpl.java]: Ack.

File: src/org/triple/banana/media/MediaYoutubeCommander.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+import org.banana.cake.interfaces.BananaTab;
+import org.triple.banana.util.UrlUtil;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public enum MediaYoutubeCommander {
+    instance;
+
+    public static final @NonNull Map<String, String> MEDIA_QUALITES = createMediaQualityMap();
```
[the remark this answers is not in the document]
Comment (jdragonbae) [reply src/org/triple/banana/media/MediaYoutubeCommander.java]: Done.

File: src/org/triple/banana/util/YouTubeUtil.java [hunk 12]
```diff
@@ -11,16 +11,116 @@
...
+        }
+
+        @NonNull
+        final ArrayList<K> getKeyList() {
+            return mList;
+        }
+    }
+
+    private static class QualityInfo {
+        private final @NonNull HashMap<String, String> mQualityMap = new HashMap<String, String>() {
```
Comment (romandev) [inline src/org/triple/banana/util/YouTubeUtil.java:73]: Shouln't be `static final`?

File: src/org/triple/banana/util/YouTubeUtil.java [hunk 13]
```diff
@@ -35,4 +45,28 @@ public static boolean isYouTubeDomainUrl(@NonNull String urlString) {
         }
         return false;
     }
+
+    public static void showQualityOptionsDialog(@Nullable Context context) {
```
[the remark this answers is not in the document]
Comment (romandev) [reply src/org/triple/banana/util/YouTubeUtil.java]: Please ignore this comment in this PR. I left this comment a long time ago but it is pushed now.

File: src/org/triple/banana/util/YouTubeUtil.java (hunk 12 above)
Comment (jdragonbae) [reply src/org/triple/banana/util/YouTubeUtil.java]: Thanks for your review! Done.

**Review by romandev — APPROVED** [review]
lgtm
